### DISCUSSION TITLE
use sum() in having() clause: coalesce nullable to bool

### DISCUSSION THREAD

**Discussion by esmoore68** [diesel-rs/diesel#3510] [Q&A] [answered]
I'm trying to write a diesel query that would create the SQL fragment:

GROUP BY player_id
HAVING SUM(achievement_value) > achievement_threshold
Here's my attempt:

```rust
           .group_by(schema::achievement::player_id)`
           .having(sum(schema::achievement::achievement_value).gt(achievement_value))
```
            

This causes the compile error:
'_^ the trait `BoolOrNullableBool` is not implemented for `(diesel::sql_types::Bool, diesel::sql_types::Nullable<diesel::sql_types::Bool>)`

I surmise that this is because the sum() function may return return null, and the having() function needs a Bool.

How can I coalesce the return value from sum to 0 to use in the having function? Or is there a better way to do what I need to do?

diesel v2.0.0
rustc: v1.67.0
Windows 11 64 bit
Postgres v14

Here is the schema for the table involved:
```rust
diesel::table! {
    achievement (id) {
        id -> Int4,
        player_id -> Varchar,
        game_code -> Varchar,
        achievement_code -> Varchar,
        achievement_value -> Float4,
        logged_on -> Timestamp,
        options -> Nullable<Jsonb>,
    }
}
```

**Comment by weiznich** [comment]
It's not possible to answer this without knowing the relevant context. Please at at least the following details to your questions:

* Information about your diesel version
* Information about your rustc version
* Information about your operating system
* Information about your database system (which one is used + what version)
* Relevant parts of your schema, code, …

**Comment by davidatsurge** [comment, marked as answer]
Indeed, like weiznich said, I'm not sure what the issue is. The following code type-checks just fine for me (using diesel 2.0.2 on nightly rust): 

```rust
use diesel::{dsl::sum, prelude::*};

mod schema {
    diesel::table! {
        achievement (id) {
            id -> Int4,
            player_id -> Varchar,
            game_code -> Varchar,
            achievement_code -> Varchar,
            achievement_value -> Float4,
            logged_on -> Timestamp,
            options -> Nullable<Jsonb>,
        }
    }
}

fn main() {
    use schema::achievement::*;

    // `rows` is Vec<String>
    let rows = table
        .select(player_id)
        .group_by(player_id)
        .having(sum(achievement_value).gt(4.0))
        .load::<String>(&mut PgConnection::establish("").unwrap())
        .unwrap();
}
```

**Reply by esmoore68** [reply]
Thank you! Your example helped to show me that it was actually my attempt to use a tuple in the having function (for multiple conditions) instead of calling .having() twice. The snippet I posted was distilled from a much more complicated query, and since I misunderstood the actual problem, I trimmed out the actual offending code. Much appreciated!
